src/sinks/datadog_archives.rs

```diff
@@ -0,0 +1,585 @@
+    /// - `date` is set from the Global Log Schema's `timestamp` mapping, or to the current time if missing;
+    /// - `message`,`host` are set from the corresponding Global Log Schema mappings;
+    /// - `source`, `service`, `status`, `tags` and other reserved attributes are left as is;
+    /// - the rest of the fields is moved to `attributes`.
+    fn encode_event(
+        &mut self,
+        event: Event,
+    ) -> Option<EncodedEvent<PartitionInnerBuffer<Vec<u8>, Bytes>>> {
+        let key = self
+            .key_prefix_template
+            .render_string(&event)
+            .map_err(|error| {
+                emit!(TemplateRenderingFailed {
+                    error,
+                    field: Some("key_prefix"),
+                    drop_event: true,
+                });
+            })
+            .ok()?;
+
+        let mut log = event.into_log();
+
+        log.insert("_id", self.generate_log_id());
+
+        let timestamp = log
+            .remove(crate::config::log_schema().timestamp_key())
+            .unwrap_or_else(|| chrono::Utc::now().timestamp_millis().into());
+        log.insert(
+            "date",
+            timestamp
+                .as_timestamp()
+                .cloned()
+                .unwrap_or_else(chrono::Utc::now)
+                .to_rfc3339_opts(SecondsFormat::Millis, true),
+        );
+        log.rename_key_flat(crate::config::log_schema().message_key(), "message");
+        log.rename_key_flat(crate::config::log_schema().host_key(), "host");
+
+        let mut attributes = BTreeMap::new();
+        let custom_attributes: Vec<String> = log
+            .keys()
+            .filter(|path| !self.reserved_attributes.contains(path.as_str()))
+            .collect();
+        for path in custom_attributes {
+            if let Some(value) = log.remove(&path) {
+                attributes.insert(path, value);
+            }
+        }
+        log.insert("attributes", attributes);
+
+        let mut bytes =
+            serde_json::to_vec(&log).expect("Failed to encode event as json, this is a bug!");
+        bytes.push(b'\n');
+
+        Some(EncodedEvent {
+            item: PartitionInnerBuffer::new(bytes, key.into()),
+            finalizers: log.metadata_mut().take_finalizers(),
+        })
+    }
+
+    /// Generates a unique ID compatible with DD
+    fn generate_log_id(&mut self) -> String {
+        let mut id = BytesMut::with_capacity(18);
+        // timestamp in millis - 6 bytes
+        let now = Utc::now();
+        id.put_int(now.timestamp_millis(), 6);
+        // namespace + version, both 0 - 2 bytes
+        id.put_i16(0);
+
+        // host-based number - 5 bytes
+        id.put_slice(&self.host_number);
+
+        // counter - 5 bytes
+        id.put_u8(0);
+        // one padding byte
+        // 4 bytes for the counter should be more than enough - it should be unique for 1 millisecond only
+        let seq_number = self.seq_number.fetch_add(1, Ordering::Relaxed);
```

Comment:
So there's a few problems with this. Firstly, relaxed ordering will mean that multiple threads would end up with the same sequence number. This will accidentally work on x86 but immediately fail on ARM. You want `AcqRel`. Second, you don't need an atomic here at all, just a plain `u32`. There's no synchronization here and no need for an atomic, which I've confirmed with the following diff:

```diff
diff --git a/src/sinks/datadog_archives.rs b/src/sinks/datadog_archives.rs
index e208a9db7..45a9f4fb9 100644
--- a/src/sinks/datadog_archives.rs
+++ b/src/sinks/datadog_archives.rs
@@ -30,7 +30,6 @@ use crate::{
     },
     template::Template,
 };
-use std::sync::atomic::{AtomicU32, Ordering};

 #[derive(Deserialize, Serialize, Debug, Clone)]
 #[serde(deny_unknown_fields)]
@@ -179,7 +178,7 @@ struct DatadogArchivesSinkEncoding {
     key_prefix_template: Template,
     reserved_attributes: HashSet<&'static str>,
     host_number: Vec<u8>,
-    seq_number: AtomicU32,
+    seq_number: u32,
 }

 impl DatadogArchivesSinkEncoding {
@@ -261,8 +260,8 @@ impl DatadogArchivesSinkEncoding {
         id.put_u8(0);
         // one padding byte
         // 4 bytes for the counter should be more than enough - it should be unique for 1 millisecond only
-        let seq_number = self.seq_number.fetch_add(1, Ordering::Relaxed);
-        id.put_u32(seq_number); // 4 bytes
+        self.seq_number = self.seq_number.wrapping_add(1);
+        id.put_u32(self.seq_number); // 4 bytes

         base64::encode(id.freeze())
     }
@@ -273,7 +272,7 @@ fn default_encoding() -> DatadogArchivesSinkEncoding {
         key_prefix_template: Template::try_from(KEY_TEMPLATE).expect("invalid object key format"),
         reserved_attributes: RESERVED_ATTRIBUTES.to_vec().into_iter().collect(),
         host_number: host_number(),
-        seq_number: AtomicU32::default(),
+        seq_number: u32::default(),
     }
 }
```

In general Rust will be _very picky_ about whether types appropriately implement [`Sync`](https://doc.rust-lang.org/std/marker/trait.Sync.html) and [`Send`](https://doc.rust-lang.org/std/marker/trait.Send.html) and you can more or less use non-atomic variants of types and expect to get dinged if you've accidentally made a goof. 

There aren't multiple threads at play here.

Reply:
oh, good to know, thanks! I didn't realise it's single-threaded, will update as you suggest.

Reply:
> oh, good to know, thanks! I didn't realise it's single-threaded, will update as you suggest.

Yep! As much as possible strive to write as if your program is single-threaded. Rust'll ding you when you violate some safety constraint.

> I thought that this will work in my case since I'm only interested in updating/reading a single value and do not care about the order of other operations.

Well, maybe that would be sufficient. If the ordering is of no consequence you're exactly right. I'd goofed and you would not have two threads getting the same sequence number. Else, the operation wouldn't be atomic. :joy:  

With regard to ordering, on a system with weak ordering -- like ARM -- it's perfectly valid for two threads to do a `fetch_add` with relaxed ordering on the same spot in memory and get a value out that is greater than the addition, from the point of view of an all-seeing observer. So, thread A does a fetch_add +1 on memory spot M with value V at time T0, thread B does a fetch_add +1 on M at time T1 > T0 but thread A gets V+2 from the operation while B gets V+1. Totally relaxed ordering allows for some really freaky shuffling by the CPU and compiler but in this case they won't manifest. The Alpha [for instance](https://en.wikipedia.org/wiki/Memory_ordering#In_symmetric_multiprocessing_(SMP)_microprocessor_systems) allowed for some very, deeply weird behavior.

Reply:
wow, that's very interesting and crazy at the same time, thanks for explaining  this 👍  I really need to deep dive into Rust's memory model and thread-safety guarantees, but really hope that I don't need to worry about things like that most of the time and can rely on the compiler :)